File: Android/app/src/main/java/com/aliyun/roompaas/app/activity/business/view/LiveInfoView.java
package com.aliyun.roompaas.app.activity.business.view;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.widget.FrameLayout;
import android.widget.TextView;

import com.aliyun.roompaas.app.R;
import com.aliyun.roompaas.app.util.AppUtil;

import java.util.Locale;

/**
 * 直播信息视图, 包含主播头像、直播标题、观看人数和点赞数信息
 *
 * @author puke
 * @version 2021/6/30
 */
public class LiveInfoView extends FrameLayout {

    private final TextView title;
    private final TextView viewCount;
    private final TextView likeCount;

    public LiveInfoView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        setMinimumHeight(AppUtil.dp(42));
        setBackgroundResource(R.drawable.bg_anchor_profile);
        inflate(context, R.layout.view_live_info, this);

        title = findViewById(R.id.view_title);
        viewCount = findViewById(R.id.view_view_count);
        likeCount = findViewById(R.id.view_like_count);
    }

    /**
     * 设置标题
     *
     * @param text 标题信息
     */
    public void setTitle(String text) {
        this.title.setText(text);
    }

    /**
     * 设置观看人数
     *
     * @param count 观看人数
     */
    public void setViewCount(int count) {
        String value = formatNumber(count);
        viewCount.setText(String.format("%s观看", value));
    }

    /**
     * 设置点赞人数
     *
     * @param count 点赞人数
     */
    public void setLikeCount(int count) {
        String value = formatNumber(count);
        likeCount.setText(String.format("%s点赞", value));
    }

    private String formatNumber(int number) {
        if ((number < 0)) {
            // 兜底保护
            return String.valueOf(0);
        } else if (number > 10000) {
            // 1w+ 格式化
            return String.format(Locale.getDefault(), "%.1fw", number / 10000f);
        } else {
            return String.valueOf(number);
        }
    }
}

File: Android/liveroom-demo/ui-ecommerce/src/main/java/com/aliyun/vpaas/standard/ecommerce/view/LiveInfoView.kt
package com.aliyun.vpaas.standard.ecommerce.view

import android.annotation.SuppressLint
import android.content.Context
import android.text.TextUtils
import android.util.AttributeSet
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.TextView
import com.alibaba.dingpaas.room.RoomDetail
import com.alibaba.dingpaas.scenelive.SceneGetLiveDetailRsp
import com.aliyun.roompaas.base.exposable.Callback
import com.aliyun.roompaas.biz.RoomEngine
import com.aliyun.roompaas.biz.SampleRoomEventHandler
import com.aliyun.roompaas.biz.exposable.event.RoomInOutEvent
import com.aliyun.roompaas.chat.SampleChatEventHandler
import com.aliyun.roompaas.chat.exposable.event.CustomMessageEvent
import com.aliyun.roompaas.roombase.Const
import com.aliyun.standard.liveroom.lib.Actions
import com.aliyun.standard.liveroom.lib.LiveContext
import com.aliyun.standard.liveroom.lib.component.BaseComponent
import com.aliyun.standard.liveroom.lib.component.ComponentHolder
import com.aliyun.standard.liveroom.lib.component.IComponent
import com.aliyun.standard.liveroom.lib.wrapper.SampleRoomEventHandlerExtends
import com.aliyun.vpaas.standard.ecommerce.R
import com.aliyun.vpaas.standard.ecommerce.custommessage.CustomMessages
import com.aliyun.vpaas.standard.ecommerce.custommessage.FollowMessage
import com.aliyun.vpaas.standard.ecommerce.model.Message
import com.bumptech.glide.Glide

/**
 * 左上角: 直播信息组件
 *
 * @author puke
 * @version 2022/5/7
 */
class LiveInfoView(context: Context, attrs: AttributeSet?) : FrameLayout(context, attrs),
    ComponentHolder {

    private val component = Component()
    private val avatar: ImageView
    private val title: TextView
    private val desc: TextView
    private val follow: TextView

    init {
        inflate(context, R.layout.live_info_view, this)
        avatar = findViewById(R.id.live_info_avatar)
        title = findViewById(R.id.live_info_title)
        desc = findViewById(R.id.live_info_desc)
        follow = findViewById(R.id.live_info_follow)
        follow.setOnClickListener {
            it.visibility = GONE
            component.handleFollow()
        }
    }

    private inner class Component : BaseComponent() {

        override fun onInit(liveContext: LiveContext?) {
            super.onInit(liveContext)
            // 主播不显示关注
            follow.visibility = if (isOwner) GONE else VISIBLE

            roomChannel.addEventHandler(object : SampleRoomEventHandlerExtends() {
                override fun onLiveRoomExtensionChanged(extension: MutableMap<String, String>?) {
                    extension?.get("anchorAvatarURL")?.run {
                        Glide.with(context).load(this).into(avatar)
                    }
                }

                override fun onEnterOrLeaveRoom(event: RoomInOutEvent?) {
                    setViewCount(event?.pv)
                }
            })

            chatService.addEventHandler(object : SampleChatEventHandler() {
                override fun onCustomMessageReceived(event: CustomMessageEvent?) {
                    CustomMessages.parseMessage(event?.data)?.run {
                        val myUserId = Const.getCurrentUserId()
                        if (this is FollowMessage && TextUtils.equals(userId, myUserId)) {
                            // 自己关注后, 做隐藏逻辑 (兼容底部点击"我也关注")
                            follow.visibility = GONE
                        }
                    }
                }
            })
        }

        override fun onEnterRoomSuccess(roomDetail: RoomDetail?) {
            setViewCount(roomDetail?.roomInfo?.pv)
        }

        @SuppressLint("SetTextI18n")
        private fun setViewCount(pv: Int?) {
            pv?.run { desc.text = "$pv 观看" }
        }

        override fun onEvent(action: String?, vararg args: Any?) {
            when (action) {
                Actions.GET_LIVE_DETAIL_SUCCESS -> {
                    liveService?.liveDetail?.liveInfo?.run {
                        this@LiveInfoView.title.text = title
                    }
                }
            }
        }

        fun handleFollow() {
            CustomMessages.doSend(chatService, FollowMessage())
        }
    }

    override fun getComponent(): IComponent {
        return component
    }
}